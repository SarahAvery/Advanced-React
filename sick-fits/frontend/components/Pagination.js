import styled from 'styled-components';
import Head from 'next/head';
import Link from 'next/link';
import gql from 'graphql-tag';
import { useQuery } from '@apollo/client';
import PaginationStyles from './styles/PaginationStyles';
import DisplayError from './ErrorMessage';
import { perPage } from '../config';

export const PAGINATION_QUERY = gql`
  query {
    _allProductsMeta {
      count
    }
  }
`;

const PaginationContainer = styled.div`
  width: 100%;
  display: flex;
  justify-content: center;
`;

// eslint-disable-next-line react/prop-types
export default function Pagination({ page }) {
  const { error, loading, data } = useQuery(PAGINATION_QUERY);
  if (loading) return <p>Loading...</p>;
  if (error) return <DisplayError error={error} />;

  const { count } = data._allProductsMeta;
  const pageCount = Math.ceil(count / perPage);

  return (
    <PaginationContainer>
      <PaginationStyles data-testid="pagination">
        <Head>
          <title>
            Sick Fits - Page {page} of {pageCount}
          </title>
        </Head>
        <Link href={`/products/${page - 1}`}>
          <a aria-disabled={page <= 1}>← Prev</a>
        </Link>
        <p>
          Page {page} of {pageCount}
        </p>
        <p>{count} Items Total</p>
        <Link href={`/products/${page + 1}`}>
          <a aria-disabled={page === pageCount}>Next →</a>
        </Link>
      </PaginationStyles>
    </PaginationContainer>
  );
}
