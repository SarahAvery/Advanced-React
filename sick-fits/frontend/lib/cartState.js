/* eslint-disable react/prop-types */
import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // this is out own custom provider. Store data (state) and functionality (updates) in here. Can be accessed via tthe consumer

  // closed by default
  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    // set the opposite of current state (true/false)
    setCartOpen(!cartOpen);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// custom hook for accessing the cart local state
function useCart() {
  // use a Consumer to access the local state
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, useCart };
