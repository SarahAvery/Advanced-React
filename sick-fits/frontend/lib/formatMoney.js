export default function formatMoney(amount = 0) {
  const options = {
    style: 'currency',
    currency: 'CAD',
    minimumFractionDidgits: 2,
  };

  if (amount % 100 === 0) {
    options.minimumFractionDidgits = 0;
  }

  const formatter = Intl.NumberFormat('en-CA', options);

  return formatter.format(amount / 100);
}
