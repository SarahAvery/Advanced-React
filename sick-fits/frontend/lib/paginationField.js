import { PAGINATION_QUERY } from '../components/Pagination';

const paginationField = () => ({
  keyArgs: false, // tells Apollo we will take care of everything
  read(existing = [], { args, cache }) {
    const { skip, first } = args;
    // Read the number of items on the page from the cache
    const data = cache.readQuery({ query: PAGINATION_QUERY });
    const count = data?._allProductsMeta?.count;
    const page = skip / first + 1;
    const pages = Math.ceil(count / first);

    // check if we have existing items
    const items = existing.slice(skip, skip + first).filter((x) => x);

    // if there are items, and there are not enough items to satisfy how many are requested
    // and on the last page
    // then just send the item(s) left (send what we have even if not full page)
    if (items.length && items.length !== first && page === pages) {
      return items;
    }
    if (items.length !== first) {
      // we dont have any items, must fetch
      return false;
    }
    //  if there are items, return them from cache
    if (items.length) {
      return items;
    }
    // fallback
    return false;
  },
  merge(existing, incoming, { args }) {
    const { skip, first } = args;
    // runs when the Apollo client comes back from the network with the products
    const merged = existing ? existing.slice(0) : [];
    // eslint-disable-next-line no-plusplus
    for (let i = skip; i < skip + incoming.length; ++i) {
      merged[i] = incoming[i - skip];
    }
    // return the merged item from the cache
    return merged;
  },
});

export default paginationField;
