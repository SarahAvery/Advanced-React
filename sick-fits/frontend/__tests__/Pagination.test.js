/* eslint-disable no-unused-vars */
import { MockedProvider } from '@apollo/client/testing';
import { render, screen } from '@testing-library/react';
import Pagination from '../components/Pagination';
import { makePaginationMocksFor } from '../lib/testUtils';

describe('<Pagination/>', () => {
  it('Displays a loading message', () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(1)}>
        <Pagination />
      </MockedProvider>
    );

    expect(container.textContent).toBe('Loading...');
    expect(container).toMatchSnapshot();
  });

  it('Renders pagination for 18 items', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );

    await screen.findByTestId('pagination');
    expect(container).toHaveTextContent('Page 1 of 9');

    expect(container).toMatchSnapshot();
  });

  it('disables the prev page on page one', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(6)}>
        <Pagination page={1} />
      </MockedProvider>
    );

    await screen.findByTestId('pagination');

    const prevButton = screen.getByText(/Prev/);
    const nextButton = screen.getByText(/Next/);

    expect(prevButton).toHaveAttribute('aria-disabled', 'true');
    expect(nextButton).toHaveAttribute('aria-disabled', 'false');
  });

  it('disables the next page on the last page', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(6)}>
        <Pagination page={3} />
      </MockedProvider>
    );

    await screen.findByTestId('pagination');

    const prevButton = screen.getByText(/Prev/);
    const nextButton = screen.getByText(/Next/);

    expect(prevButton).toHaveAttribute('aria-disabled', 'false');
    expect(nextButton).toHaveAttribute('aria-disabled', 'true');
  });

  it('Enables all on middle page', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={makePaginationMocksFor(6)}>
        <Pagination page={2} />
      </MockedProvider>
    );

    await screen.findByTestId('pagination');

    const prevButton = screen.getByText(/Prev/);
    const nextButton = screen.getByText(/Next/);

    expect(prevButton).toHaveAttribute('aria-disabled', 'false');
    expect(nextButton).toHaveAttribute('aria-disabled', 'false');
  });
});
