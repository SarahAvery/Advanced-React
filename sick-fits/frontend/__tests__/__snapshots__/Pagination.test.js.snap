// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> Displays a loading message 1`] = `
<div>
  <p>
    Loading...
  </p>
</div>
`;

exports[`<Pagination/> Renders pagination for 18 items 1`] = `
<div>
  <div
    class="Pagination__PaginationContainer-sc-w8etnd-0 gzbFKH"
  >
    <div
      class="PaginationStyles-sc-aduuar-0 kNLCHA"
      data-testid="pagination"
    >
      <a
        aria-disabled="true"
        href="/products/0"
      >
        ← Prev
      </a>
      <p>
        Page 
        1
         of 
        9
      </p>
      <p>
        18
         Items Total
      </p>
      <a
        aria-disabled="false"
        href="/products/2"
      >
        Next →
      </a>
    </div>
  </div>
</div>
`;
